# Praktikum Kecerdasan Buatan Naïve Bayes

| | |
|---|---|
| **Nama** | Laura Aurelia |
| **NIM** | 42324027 |
| **Prodi** | D3 Teknologi Informasi |
| **Mata Kuliah** | 4233201 – Kecerdasan Buatan |
| **Topik** | Naïve Bayes |

Naïve Bayes

Naïve Bayes adalah algoritma klasifikasi berbasis probabilitas yang menggunakan Teorema Bayes, dengan asumsi bahwa semua fitur dalam dataset bersifat independent atau saling independen.

Teorema Bayes

Rumus Teorema Bayes:

$$ P(C|X) = \frac{P(X|C)P(C)}{P(X)} $$

Keterangan:

P(C|X) : Probabilitas kelas C jika diberikan fitur X
P(X|C) : Probabilitas fitur X jika kelasnya C
P(C) : Probabilitas awal suatu kelas
P(X) : Probabilitas fitur X pada seluruh data

Terdapat tiga jenis utama Naïve Bayes, yaitu Gaussian Naïve Bayes, Multinomial Naïve Bayes, dan Bernoulli Naïve Bayes. Jenis yang digunakan disesuaikan dengan bentuk data yang digunakan.

1. Gaussian Naïve Bayes

Gaussian Naïve Bayes digunakan ketika fitur dalam dataset berupa data numerik kontinu dan diasumsikan mengikuti distribusi normal atau Gaussian.

$$ P(X|C) = \frac{1}{\sigma\sqrt{2\pi}} e^{-\frac{(X-\mu)^2}{2\sigma^2}} $$

Keterangan:

μ : Rata-rata fitur dalam kelas tertentu
σ : Standar deviasi fitur dalam kelas tertentu
2. Multinomial Naïve Bayes

2. Multinomial Naïve Bayes digunakan untuk data kategori yang berdasarkan frekuensi kemunculan suatu fitur.

$$ P(X|C) = \frac{count(X,C)+\alpha} {\sum count(X',C)+\alpha V} $$

Keterangan:

count(X,C) : Jumlah kemunculan fitur X dalam kelas C
count(X',C) : Total jumlah semua kata dalam kelas tertentu
V : Jumlah fitur unik
α : Laplace Smoothing, biasanya bernilai 1 untuk menghindari probabilitas nol
3. Bernoulli Naïve Bayes

3. Bernoulli Naïve Bayes digunakan untuk data biner (0/1). Pada jenis ini, fitur menunjukkan apakah suatu elemen ada atau tidak ada dalam suatu kategori.

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris

# Load dataset Iris dari scikit-Learn
iris = load_iris()

Pada tahap pertama, kita memasukkan library yang dibutuhkan untuk praktikum. numpy digunakan untuk mengolah data dalam bentuk array, sedangkan pandas digunakan untuk mengolah data dalam bentuk tabel.

Selanjutnya, load_iris digunakan untuk mengambil dataset Iris yang sudah disediakan oleh scikit-learn. Dataset ini nantinya akan digunakan sebagai data yang akan diklasifikasikan.

In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(data=iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]

# Defenisikan  X (Fitur) dan Y (Target)
X = df.drop(columns=["Species"])
Y = df["Species"]

print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


Pada bagian ini, data Iris diubah menjadi bentuk DataFrame agar lebih mudah digunakan dan diolah. Data tersebut berisi empat fitur bunga, yaitu sepal length, sepal width, petal length, dan petal width. Kemudian ditambahkan kolom Species yang berisi jenis bunga Iris, yaitu setosa, versicolor, dan virginica. Setelah itu, data dipisahkan menjadi X sebagai fitur atau ciri-ciri bunga yang digunakan untuk melakukan prediksi, sedangkan Y sebagai target atau jenis bunga yang akan diprediksi. print(X) digunakan untuk menampilkan data fitur yang akan digunakan dalam proses klasifikasi.

In [3]:
Y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

In [5]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(Y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Pada bagian ini, LabelEncoder digunakan untuk mengubah label jenis bunga yang masih berbentuk teks menjadi angka agar dapat diproses oleh algoritma Naïve Bayes. Jadi, label seperti setosa, versicolor, dan virginica akan diubah menjadi angka, yaitu 0, 1, dan 2. Hasil perubahan ini nantinya digunakan sebagai target (y) dalam proses pelatihan dan pengujian model.

In [6]:
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

Pada bagian ini, dataset dibagi menjadi data training dan data testing. test_size=0.2 berarti 20% data digunakan untuk testing, sedangkan 80% sisanya digunakan untuk training. random_state=0 digunakan supaya pembagian data tetap sama setiap kali program dijalankan. Jadi, X_train dan y_train digunakan untuk melatih model, sedangkan X_test dan y_test digunakan untuk menguji hasil prediksi model.

In [7]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
NB_model = model.fit(X_train, y_train)
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

Karena fitur pada dataset Iris berupa data numerik kontinu, maka digunakan Gaussian Naïve Bayes. Model dibuat menggunakan GaussianNB(), kemudian dilatih menggunakan data training (X_train dan y_train). Setelah model selesai dilatih, model digunakan untuk memprediksi data testing (X_test). Hasil prediksi tersebut disimpan dalam result.

In [8]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


Pada bagian ini, kita menghitung tingkat akurasi model berdasarkan hasil prediksi pada data testing. accuracy_score() membandingkan hasil sebenarnya (y_test) dengan hasil prediksi model (result). Hasilnya kemudian dikalikan 100 agar ditampilkan dalam bentuk persentase. Pada praktikum ini, diperoleh accuracy sebesar 96.66666666666667%, yang berarti model berhasil memprediksi sebagian besar data testing dengan benar.

In [9]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


C:\Users\ASUS\AppData\Roaming\Python\Python314\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


Pada bagian ini, model digunakan untuk memprediksi data bunga baru dengan nilai [5.1, 3.5, 1.4, 0.2]. Data tersebut dimasukkan ke dalam model yang sudah dilatih sebelumnya. Hasil prediksi awal masih berupa angka karena sebelumnya label telah diubah menggunakan LabelEncoder. Oleh karena itu, inverse_transform() digunakan untuk mengubah kembali angka hasil prediksi menjadi nama species. Hasilnya adalah setosa.

Ketika kode sebelumnya dijalankan, muncul warning karena model dilatih menggunakan DataFrame yang memiliki nama kolom, sedangkan data baru diberikan langsung dalam bentuk list/array sehingga tidak memiliki nama kolom.

Solusinya, data baru dibuat menjadi DataFrame dengan nama kolom yang sama seperti dataset Iris.

In [10]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


Dengan cara ini, struktur data yang diberikan saat prediksi menjadi konsisten dengan struktur data yang digunakan ketika model dilatih. Hasil prediksinya tetap setosa, tetapi warning sebelumnya tidak muncul lagi.

Tugas!
1. Ikuti semua tahapan dalam modul ini.
2. Prediksi data baru masuk ke species apa: [7.3, 3.8, 4.6, 2.3]

In [11]:
input_data = pd.DataFrame(np.array([[7.3, 3.8, 4.6, 2.3]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['virginica']


Pada praktikum ini, saya belajar menggunakan algoritma Naïve Bayes untuk mengklasifikasikan data pada dataset Iris. Data terlebih dahulu dipisahkan menjadi data training dan testing, kemudian digunakan untuk membuat model Gaussian Naïve Bayes karena fitur pada dataset berupa data numerik. Dari hasil pengujian, model mendapatkan accuracy sebesar 96,67%. Model juga dapat digunakan untuk memprediksi data baru. Pada tugas yang diberikan, data [7.3, 3.8, 4.6, 2.3] diprediksi termasuk ke dalam species virginica.